include_physics/ideal_MHD.h

```diff
@@ -388,12 +388,177 @@ inline void init_fields(uint ix, uint iy, uint iz, global REAL* u) {
 #ifndef USE_PERIODIC
-  #error "Error in ideal_MHD.cl: Surface terms not implemented for nonperiodic boundaries!"
-#endif // USE_PERIODIC*/
+  int i;
+  REAL um[NUM_TOTAL_VARS] = {0.0};
+  get_field(ix, iy, iz, 0, 0, 0, u, um); 
+  REAL4 b_bound = b_boundary(ix, iy, iz, time);
+  REAL4 u_bound = u_boundary(ix, iy, iz, time);
+  REAL rho_bound = (REAL)1;
+  REAL p_bound = 1;
+  REAL E_bound = compute_energy(p_bound, rho_bound, u_bound.x, u_bound.y, u_bound.z, b_bound.x, b_bound.y, b_bound.z);
+
+
+  REAL ub[NUM_TOTAL_VARS] = {0.0};
+  ub[Field_ux] = u_bound.x;
+  ub[Field_uy] = u_bound.y;
+  ub[Field_uz] = u_bound.z;
+  ub[Field_rho_ux] = u_bound.x * rho_bound;
+  ub[Field_rho_uy] = u_bound.y * rho_bound;
+  ub[Field_rho_uz] = u_bound.z * rho_bound;
+  ub[Field_rho] = rho_bound;
+  ub[Field_p] = p_bound;
+  ub[Field_E] = E_bound;
+  ub[Field_Bx] = b_bound.x;
+  ub[Field_By] = b_bound.y;
+  ub[Field_Bz] = b_bound.z;
+
+  REAL flux[NUM_CONSERVED_VARS] = {0.0};
+  REAL fluxm[NUM_CONSERVED_VARS] = {0.0};
+  REAL fluxb[NUM_CONSERVED_VARS] = {0.0};
+
+  REAL BBSQb = pow(ub[Field_Bx], 2) + pow(ub[Field_By], 2) + pow(ub[Field_Bz], 2);
+  REAL BBSQm = pow(um[Field_Bx], 2) + pow(um[Field_By], 2) + pow(um[Field_Bz], 2);
+
+
+  // X direction
+  REAL cfbx = sqrt((GAMMA * ub[Field_p] + BBSQb + sqrt(pow(GAMMA * ub[Field_p] + BBSQb, 2) - 4 * GAMMA * ub[Field_p] * ub[Field_Bx] * ub[Field_Bx]))/(2 * ub[Field_rho]));
+
+  REAL cfmx = sqrt((GAMMA * um[Field_p] + BBSQm + sqrt(pow(GAMMA * um[Field_p] + BBSQm, 2) - 4 * GAMMA * um[Field_p] * um[Field_Bx] * um[Field_Bx]))/(2 * um[Field_rho]));
+
+  // Y direction
+  REAL cfby = sqrt((GAMMA * ub[Field_p] + BBSQb + sqrt(pow(GAMMA * ub[Field_p] + BBSQb, 2) - 4 * GAMMA * ub[Field_p] * ub[Field_By] * ub[Field_By]))/(2 * ub[Field_rho]));
+
+  REAL cfmy = sqrt((GAMMA * um[Field_p] + BBSQm + sqrt(pow(GAMMA * um[Field_p] + BBSQm, 2) - 4 * GAMMA * um[Field_p] * um[Field_By] * um[Field_By]))/(2 * um[Field_rho]));
+
+  // Z direction
+  REAL cfbz = sqrt((GAMMA * ub[Field_p] + BBSQb + sqrt(pow(GAMMA * ub[Field_p] + BBSQb, 2) - 4 * GAMMA * ub[Field_p] * ub[Field_Bz] * ub[Field_Bz]))/(2 * ub[Field_rho]));
+
+  REAL cfmz = sqrt((GAMMA * um[Field_p] + BBSQm + sqrt(pow(GAMMA * um[Field_p] + BBSQm, 2) - 4 * GAMMA * um[Field_p] * um[Field_Bz] * um[Field_Bz]))/(2 * um[Field_rho]));
+
+
+  REAL alx = fmin(ub[Field_ux] - cfbx, um[Field_ux] - cfmx);
+  REAL arx = fmax(ub[Field_ux] + cfbx, um[Field_ux] + cfmx);
+ 
+  REAL aly = fmin(ub[Field_uy] - cfby, um[Field_uy] - cfmy);
+  REAL ary = fmax(ub[Field_uy] + cfby, um[Field_uy] + cfmy);
+
+  REAL alz = fmin(ub[Field_uz] - cfbz, um[Field_uz] - cfmz);
+  REAL arz = fmax(ub[Field_uz] + cfbz, um[Field_uz] + cfmz);
+
+
+
+  if (check_bound_xr(ix, 1)){
+	  calc_flux_f(um, fluxm);
+	  calc_flux_f(ub, fluxb);
+          calc_num_flux(alx, arx, um, ub, fluxm, fluxb, flux);
+          for(i = 0; i < NUM_CONSERVED_VARS; i++)
+          	du_dt[i] -= M_INV[0] / DX * (flux[i] - fluxm[i]);
```

Comment:
Interestingly, adding braces around `M_INV[0] / DX` improves the performance a bit on my system. Maybe the compiler is able to figure out that the division needs to be computed only once in that case?